reflect_struct_decoder.go

```diff
@@ -32,9 +32,12 @@ func decoderOfStruct(ctx *ctx, typ reflect2.Type) ValDecoder {
 	for k, binding := range bindings {
 		fields[k] = binding.Decoder.(*structFieldDecoder)
 	}
-	for k, binding := range bindings {
-		if _, found := fields[strings.ToLower(k)]; !found {
-			fields[strings.ToLower(k)] = binding.Decoder.(*structFieldDecoder)
+
+	if !ctx.caseSensitive() {
+		for k, binding := range bindings {
+			if _, found := fields[strings.ToLower(k)]; !found {
+				fields[strings.ToLower(k)] = binding.Decoder.(*structFieldDecoder)
```

Comment:
Do the toLower calls on lines 521 and 527 need conditionalizing as well

Reply:
And if so, is there a test case that could exercise that path?

Reply:
> Do the toLower calls on lines 521 and 527 need conditionalizing as well?

I don't think so. In fact, I don't think we even need to check for the nil condition there.

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_struct_decoder.go#L516-L519

Here `decoder` is a `generalStructDecoder`, which is created here:

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_struct_decoder.go#L485

The `fields` in this struct comes from:

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_struct_decoder.go#L31-L34

The `*structFieldDecoder` is a `*StructDescriptor` which is created from:

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_struct_decoder.go#L14

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_extension.go#L324

which creates from:

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_extension.go#L388-L391

From the above snippet ^, `*StructDescriptor` can never be nil.

Also, if the decoder is supposed to be case sensitive, the conversion to lower case is take care of while constructing the struct decoder:

https://github.com/json-iterator/go/blob/7cceb6c2e37f896dc3817e031f83dce25b180a8b/reflect_struct_decoder.go#L35-L39

so the `toLower()` which is used below in lines 521 and 527 is not called and not needed.

Reply:
If this makes sense, I'll remove the lines. But I want to double check if this is indeed correct first.

Reply:
> Also, if the decoder is supposed to be case sensitive, the conversion to lower case is take care of while constructing the struct decoder:

That's my understanding as well. I agree that we should remove the lines.

Reply:
> That's my understanding as well. I agree that we should remove the lines.

Removed. Ptal, thanks.

Reply:
Wait, if the decoder is case insensitive, we need to keep the lines, because `decoder.fields[field]` might return empty, and we can only find the decoder via `decoder.fields[strings.ToLower(field)]`.

Reply:
> Yup, added back the lines along with a new test case.

+1